zerocopy-derive/src/lib.rs

```diff
@@ -1335,18 +1429,45 @@ fn impl_block<D: DataExt>(
         }
     });
 
+    let outer_extras = if let Some(repr) = known_layout_maybe_uninit_struct_repr {
+        // PANICS: TODO
+        let (trailing_field, leading_fields) = fields.split_last().unwrap();
+        let (_name, trailing_field_ty) = trailing_field;
+        let leading_fields_tys = leading_fields.iter().map(|(_name, ty)| ty);
+        let params = params.clone();
+        let bounds = bounds.clone();
+        quote! {
+            #repr
+            #[doc(hidden)]
+            pub struct __ZerocopyKnownLayoutMaybeUninit<#(#params),*>(
+                #(
+                    ::zerocopy::util::macro_util::core_reexport::mem::MaybeUninit<#leading_fields_tys>,
+                )*
+                <#trailing_field_ty as ::zerocopy::KnownLayout>::MaybeUninit
+            ) where #(#bounds,)*;
+        }
+    } else {
+        quote! {}
+    };
```

Comment:
This is all a hack - this should live in `fn derive_known_layout`. It repeats a lot of the logic from that function.